File: tools/sources/java/Exceptions.java
// Java 21 exception handling

import java.io.*;

public class Exceptions {

    // Basic try-catch
    public static void basicException() {
        try {
            int result = 10 / 0;
        } catch (ArithmeticException e) {
            System.out.println("Division by zero");
        }
    }

    // Multiple catch blocks
    public static void multipleCatch() {
        try {
            String str = null;
            int length = str.length();
        } catch (NullPointerException e) {
            System.out.println("Null pointer");
        } catch (IllegalArgumentException e) {
            System.out.println("Illegal argument");
        }
    }

    // Multi-catch (Java 7+)
    public static void multiCatch() {
        try {
            // Some code
        } catch (IOException | SQLException e) {
            System.out.println("IO or SQL exception");
        }
    }

    // Try-with-finally
    public static void withFinally() {
        try {
            riskyOperation();
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            System.out.println("Always executed");
        }
    }

    // Try-with-resources (Java 7+)
    public static void tryWithResources() {
        try (BufferedReader reader = new BufferedReader(new FileReader("file.txt"))) {
            String line = reader.readLine();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Multiple resources in try-with-resources
    public static void multipleResources() {
        try (FileInputStream fis = new FileInputStream("input.txt");
             FileOutputStream fos = new FileOutputStream("output.txt")) {
            // Use streams
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Custom exception
    public static class ValidationException extends Exception {
        public ValidationException(String message) {
            super(message);
        }

        public ValidationException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    // Custom runtime exception
    public static class ConfigurationException extends RuntimeException {
        public ConfigurationException(String message) {
            super(message);
        }
    }

    // Throwing exceptions
    public static void validateAge(int age) throws ValidationException {
        if (age < 0) {
            throw new ValidationException("Age cannot be negative");
        }
        if (age > 150) {
            throw new ValidationException("Age is too high");
        }
    }

    // Rethrowing exceptions
    public static void rethrowExample() throws IOException {
        try {
            readFile();
        } catch (IOException e) {
            System.out.println("Logging error");
            throw e; // Rethrow
        }
    }

    // Exception chaining
    public static void chainedException() {
        try {
            throw new IOException("Original error");
        } catch (IOException e) {
            throw new RuntimeException("Wrapped error", e);
        }
    }

    // Suppressed exceptions (try-with-resources)
    public static void suppressedExceptions() {
        try (Resource r = new Resource()) {
            throw new Exception("Primary exception");
        } catch (Exception e) {
            Throwable[] suppressed = e.getSuppressed();
        }
    }

    static class Resource implements AutoCloseable {
        @Override
        public void close() throws Exception {
            throw new Exception("Close exception");
        }
    }

    // Checked vs unchecked
    public static void checkedVsUnchecked() throws IOException {
        // Checked exception - must be declared or caught
        throw new IOException("Checked");

        // Unchecked exception - no declaration needed
        // throw new RuntimeException("Unchecked");
    }

    // Finally without catch
    public static void finallyWithoutCatch() {
        try {
            riskyOperation();
        } finally {
            cleanup();
        }
    }

    // Nested try-catch
    public static void nestedTryCatch() {
        try {
            try {
                int result = 10 / 0;
            } catch (ArithmeticException e) {
                throw new RuntimeException("Converted", e);
            }
        } catch (RuntimeException e) {
            System.out.println("Outer catch");
        }
    }

    private static void riskyOperation() throws Exception {
        throw new Exception("Something went wrong");
    }

    private static void readFile() throws IOException {
        throw new IOException("File not found");
    }

    private static void cleanup() {
        System.out.println("Cleaning up");
    }
}

File: tools/sources/python/exceptions.py
"""Python 3 exception handling."""


def basic_exception_handling():
    """Basic try/except."""
    try:
        result = 10 / 0
    except ZeroDivisionError:
        print("Cannot divide by zero")


def multiple_exceptions():
    """Handling multiple exception types."""
    try:
        value = int("not a number")
    except ValueError:
        print("Invalid value")
    except TypeError:
        print("Invalid type")


def exception_with_else():
    """Try/except with else clause."""
    try:
        result = 10 / 2
    except ZeroDivisionError:
        print("Error")
    else:
        print(f"Result: {result}")


def exception_with_finally():
    """Try/except with finally."""
    try:
        file = open('test.txt')
        data = file.read()
    except FileNotFoundError:
        print("File not found")
    finally:
        print("Cleanup")


def catch_all_exceptions():
    """Catching all exceptions."""
    try:
        risky_operation()
    except Exception as ee:
        print(f"Error: {ee}")


def risky_operation():
    raise ValueError("Something went wrong")


class ValidationError(Exception):
    """Custom exception class."""

    def __init__(self, field, message):
        self.field = field
        self.message = message
        super().__init__(f"{field}: {message}")


class DatabaseError(Exception):
    """Custom exception with additional attributes."""

    def __init__(self, query, error_code):
        self.query = query
        self.error_code = error_code
        super().__init__(f"Query failed: {query}")


def raise_custom_exception():
    """Raising custom exceptions."""
    raise ValidationError("email", "Invalid format")


def reraise_exception():
    """Re-raising exceptions."""
    try:
        raise ValueError("Original error")
    except ValueError:
        print("Logging error")
        raise


def exception_chaining():
    """Exception chaining with from."""
    try:
        int("invalid")
    except ValueError as e:
        raise RuntimeError("Processing failed") from e


def suppress_exception():
    """Suppressing exceptions."""
    try:
        raise ValueError("Error")
    except ValueError:
        pass
